File: Dataset/Concurrency_prepared/Java/1117_trace_4.java
class H2O {
    private Semaphore h = new Semaphore(2);
    private Semaphore o = new Semaphore(0);

    public H2O() {
    }

    public void hydrogen(Runnable releaseHydrogen) throws InterruptedException {
        h.acquire();
        // releaseHydrogen.run() outputs "H". Do not change or remove this line.
        releaseHydrogen.run();
        o.release();
    }

    public void oxygen(Runnable releaseOxygen) throws InterruptedException {
        o.acquire(2);
        // releaseOxygen.run() outputs "O". Do not change or remove this line.
        releaseOxygen.run();
        h.release(2);
    }

    public static void main(String[] args) throws InterruptedException {
        H2O h2o = new H2O();
        Runnable releaseHydrogen = () -> System.out.print("H");
        Runnable releaseOxygen = () -> System.out.print("O");
        h2o.hydrogen(releaseHydrogen);
        h2o.oxygen(releaseOxygen);
    }
}

File: Dataset/Concurrency_prepared/Python/1117_trace_4.py
from threading import Semaphore


class H2O:
    def __init__(self):
        self.h = Semaphore(2)
        self.o = Semaphore(0)

    def hydrogen(self, releaseHydrogen: "Callable[[], None]") -> None:
        self.h.acquire()
        # releaseHydrogen() outputs "H". Do not change or remove this line.
        releaseHydrogen()
        if self.h._value == 0: # semaphore value checker api
            self.o.release()

    def oxygen(self, releaseOxygen: "Callable[[], None]") -> None:
        self.o.acquire()
        # releaseOxygen() outputs "O". Do not change or remove this line.
        releaseOxygen()
        self.h.release(2)


if __name__ == "__main__":
    import threading
    import time

    def releaseHydrogen():
        print("H", end="")

    def releaseOxygen():
        print("O", end="")

    h2o = H2O()

    threads = []
    for _ in range(8):
        threads.append(threading.Thread(target=h2o.hydrogen, args=(releaseHydrogen,)))
    for _ in range(4):
        threads.append(threading.Thread(target=h2o.oxygen, args=(releaseOxygen,)))

    for t in threads:
        t.start()
        #START
    for t in threads:
        t.join()